### PR TITLE
Limit parallel TLS connections

### LINKED ISSUE #6517: Not all Endpoints can't reconnect due to "Client TLS handshake failed" error after "reload or restart"
Not all Endpoints can't reconnect due to "Client TLS handshake failed" error after "reload or restart"

After upgrading to Icinga-2.9.0 and 2.9.1 we ran into a huge problem with reconnecting to our endpoints.

Do a "systemctl restart icinga2", a "systemctl reload icinga2" or a reload in Icingaweb2 and Director-1.4.3.
The following is happening:
```
[2018-08-02 15:29:29 +0200] information/ExternalCommandListener: 'command' stopped.
[2018-08-02 15:31:10 +0200] information/WorkQueue: #10 (DaemonCommand::Run) items: 0, rate:  0/s (0/min 0/5min 0/15min);
[2018-08-02 15:31:12 +0200] information/ApiListener: 'api' started.
[2018-08-02 15:31:12 +0200] information/ApiListener: Copying 1 zone configuration files for zone 'mon-icinga2-02.nst.example.org' to '/var/lib/icinga2/api/zones/mon-icinga2-02.nst.example.org'.
[2018-08-02 15:31:12 +0200] information/ApiListener: Applying configuration file update for path '/var/lib/icinga2/api/zones/mon-icinga2-02.nst.example.org' (0 Bytes). Received timestamp '2018-08-02 15:31:12 +0200' (1533216672.088238), Current timestamp '2018-08-02 13:32:44 +0200' (1533209564.472608).
[2018-08-02 15:31:12 +0200] information/ApiListener: Copying 1 zone configuration files for zone 'mon-icinga2-03.nst.example.org' to '/var/lib/icinga2/api/zones/mon-icinga2-03.nst.example.org'.
[2018-08-02 15:31:12 +0200] information/ApiListener: Applying configuration file update for path '/var/lib/icinga2/api/zones/mon-icinga2-03.nst.example.org' (0 Bytes). Received timestamp '2018-08-02 15:31:12 +0200' (1533216672.097991), Current timestamp '2018-08-02 13:32:44 +0200' (1533209564.484660).
[2018-08-02 15:31:12 +0200] information/ApiListener: Copying 10 zone configuration files for zone 'mon-icinga2-01.nst.example.org' to '/var/lib/icinga2/api/zones/mon-icinga-01.nst.example.org'.
[2018-08-02 15:31:12 +0200] information/ApiListener: Applying configuration file update for path '/var/lib/icinga2/api/zones/mon-icinga-01.example.org' (0 Bytes). Received timestamp '2018-08-02 15:31:12 +0200' (1533216672.178659), Current timestamp '2018-08-02 13:32:44 +0200' (1533209564.565199).
[2018-08-02 15:31:12 +0200] information/ApiListener: Copying 12 zone configuration files for zone 'director-global' to '/var/lib/icinga2/api/zones/director-global'.
[2018-08-02 15:31:12 +0200] information/ApiListener: Applying configuration file update for path '/var/lib/icinga2/api/zones/director-global' (0 Bytes). Received timestamp '2018-08-02 15:31:12 +0200' (1533216672.184592), Current timestamp '2018-08-02 13:32:44 +0200' (1533209564.571902).
[2018-08-02 15:31:12 +0200] information/ApiListener: Adding new listener on port '5665'
[2018-08-02 15:31:12 +0200] information/ApiListener: Reconnecting to endpoint 'system1.example.org' via host '10.0.1.104' and port '5665'
[2018-08-02 15:31:12 +0200] information/ApiListener: Reconnecting to endpoint 'system2.example.org' via host '10.0.2.104' and port '5665'
[2018-08-02 15:31:12 +0200] information/ApiListener: Reconnecting to endpoint 'system3.example.org' via host '10.0.3.104' and port '5665'
...
[2018-08-02 15:31:14 +0200] information/ApiListener: Reconnecting to endpoint 'system1134.example.org' via host '10.5.1.101' and port '5665'
...
```
Between 15:31:12 and 15:31:14 a reconnect is triggered (and logged) for all 6897 endpoints.

For a few endpoint this is successful, the connection is established and config files are synced and updated.

```
[2018-08-02 15:31:22 +0200] information/ApiListener: Finished sending config file updates for endpoint 'system5.example.org' in zone 'system5.example.org'.
[2018-08-02 15:31:22 +0200] information/ApiListener: Syncing runtime objects to endpoint 'system5.example.org'.
```

Suddenly the CLient TLS handshakes appear

```
[2018-08-02 15:31:22 +0200] critical/ApiListener: Client TLS handshake failed (to [10.3.2.103]:5665)
Context:
        (0) Handling new API client connection

[2018-08-02 15:31:22 +0200] critical/ApiListener: Client TLS handshake failed (to [10.1.27.154]:5665)
Context:
        (0) Handling new API client connection

[2018-08-02 15:31:22 +0200] critical/ApiListener: Client TLS handshake failed (to [10.245.10.103]:5665)
Context:
        (0) Handling new API client connection
...
[2018-08-02 15:31:22 +0200] critical/ApiListener: Client TLS handshake failed (to [10.127.11.99]:5665)
Context:
        (0) Handling new API client connection

```

On client side (10.127.11.99) the following will occur in the logfile

```
...
[2018-08-02 15:31:23 +0200] critical/ApiListener: Client TLS handshake failed (from [10.195.1.10]:33143)
Context:
        (0) Handling new API client connection

[2018-08-02 15:31:31 +0200] information/JsonRpcConnection: No messages for identity 'mon-icinga2-01.nst.example.org' have been received in the last 60 seconds.
[2018-08-02 15:31:31 +0200] warning/JsonRpcConnection: API client disconnected for identity 'mon-icinga2-01.nst.example.org'
[2018-08-02 15:31:31 +0200] warning/JsonRpcConnection: API client disconnected for identity 'mon-icinga2-01.nst.example.org'
[2018-08-02 15:31:31 +0200] warning/ApiListener: Removing API client for endpoint 'mon-icinga2-01.nst.example.org'. 0 API clients left.
...
[2018-08-02 15:32:24 +0200] critical/ApiListener: Client TLS handshake failed (from [10.195.1.10]:39162)
Context:
        (0) Handling new API client connection

[2018-08-02 15:32:44 +0200] information/ConfigObject: Dumping program state to file '/var/lib/icinga2/icinga2.state'
[2018-08-02 15:33:25 +0200] critical/ApiListener: Client TLS handshake failed (from [10.195.1.10]:44709)
Context:
        (0) Handling new API client connection

[2018-08-02 15:34:25 +0200] information/ApiListener: New client connection for identity 'mon-icinga2-01.nst.example.org' from [10.195.1.10]:51952
[2018-08-02 15:34:30 +0200] warning/TlsStream: TLS stream was disconnected.
[2018-08-02 15:34:30 +0200] warning/ApiListener: No data received on new API connection for identity 'mon-icinga2-01.nst.example.org'. Ensure that the remote endpoints are properly configured in a cluster setup.
Context:
        (0) Handling new API client connection

[2018-08-02 15:35:13 +0200] information/WorkQueue: #6 (ApiListener, SyncQueue) items: 0, rate:  0/s (0/min 0/5min 0/15min);

```

Also the API gets unresponsive now and you can't query something. System and database load isn't higher as in 2.8.4 at this time.

A few minutes later, some system succeed in reconneting and syncing, other not.

```
[2018-08-02 15:45:39 +0200] information/ApiListener: Finished reconnecting to endpoint 'system479.example.org' via host '10.52.217.105' and port '5665'
[2018-08-02 15:45:39 +0200] information/ApiListener: Finished reconnecting to endpoint 'system325.example.org' via host '10.71.10.158' and port '5665'
[2018-08-02 15:45:39 +0200] critical/ApiListener: Client TLS handshake failed (to [10.5.169.2]:5665)
Context:
        (0) Handling new API client connection

[2018-08-02 15:45:39 +0200] critical/ApiListener: Client TLS handshake failed (to [10.11.1.3]:5665)
Context:
        (0) Handling new API client connection

[2018-08-02 15:45:39 +0200] critical/ApiListener: Client TLS handshake failed (to [10.8.64.43]:5665)
Context:
        (0) Handling new API client connection

[2018-08-02 15:45:39 +0200] information/ApiListener: Finished reconnecting to endpoint 'system123.example.org' via host '10.25.78.109' and port '5665'
[2018-08-02 15:45:39 +0200] information/ApiListener: Finished reconnecting to endpoint 'system5.example.org' via host '10.61.80.159' and port '5665'
[2018-08-02 15:45:39 +0200] information/ApiListener: Finished reconnecting to endpoint 'system12.example.org' via host '10.25.149.110' and port '5665'
```

Another few minutes later, about 1000 systems could do a successful reconnect without any TLS errors, but then TLS errors in logfile raises again. In half an hour there are only ~ 3000 endpoints reconnected, all others have to deal with the strange TLS handshake failures

```
[2018-08-02 15:31:23 +0200] critical/ApiListener: Client TLS handshake failed (from [10.195.1.10]:33143)
Context:
        (0) Handling new API client connection

[2018-08-02 15:31:31 +0200] information/JsonRpcConnection: No messages for identity 'mon-icinga-01.nst.example.org' have been received in the last 60 seconds.
```

With 2.9.0 and 2.9.1 it isn't possible to reconnect all endpoints in a time effective manner.

A Rollback to 2.8.4 fixes this problem immediately. After approximately 5 minutes all endpoints are reconnected.

## Expected Behavior
Reconnection in 2.9.0 and 2.9.1 should work as in 2.8.x and before.

## Current Behavior
It's not possible to do reconnect for ALL endpoints. Only a few ( ~ 1500 till 2000) could do a successful reconnect. All others have to deal with TLS errors nd possible timeouts.

Additionally a restart or reload creates a process in <defunct> state

```
root     22070   823  0 15:30 ?        00:00:00 [icinga2] <defunct>
root     22257   823  0 12:02 ?        00:00:00 [icinga2] <defunct>
```

```
[2018-08-02 15:31:31 +0200] information/JsonRpcConnection: No messages for identity 'mon-icinga-01.nst.example.org' have been received in the last 60 seconds.
```
## Possible Solution
Rollback to 2.8.4 at the moment ...

## Steps to Reproduce (for bugs)
Hard to reproduce, because you need at least ~ 7000 endpoints and ~ 140.000 services connected.

## Your Environment
* Version used (`icinga2 --version`):

Rollback done to 2.8.4-1

```
icinga2 - The Icinga 2 network monitoring daemon (version: r2.8.4-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Debian GNU/Linux
  Platform version: 8 (jessie)
  Kernel: Linux
  Kernel version: 3.16.0-6-amd64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.9.2
  Build host: 289f70f9cece
```
* Operating System and version:
    * Debian 8, x86_64

* Enabled features (`icinga2 feature list`):

** Disabled features: compatlog debuglog elasticsearch gelf graphite livestatus opentsdb perfdata statusdata syslog
    * Enabled features: api checker command ido-mysql influxdb mainlog notification

* Icinga Web 2 version and modules (System - About):
* Config validation (`icinga2 daemon -C`):
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.

    * icinga2 object list --type Zone | grep Object | wc -l
q6898

    * icinga2 object list --type Zone | grep Object | wc -l
6899

### PR DESCRIPTION
**Pull request by N-o-X**
# Tests

- [x] Master connects to many clients (socat)
- [x] Master connects to many clients (Icinga instances)
- [ ] Master connects, firewall blocks any request
- [x] Many clients connect to master
- [x] Many REST API clients
- [ ] Master => clients & clients => master & firewall & REST API

fixes #6517

### CODE REVIEW

File: lib/remote/apilistener.cpp [hunk 1]
```diff
@@ -43,17 +45,52 @@ REGISTER_TYPE(ApiListener);
...
 	m_RelayQueue.SetName("ApiListener, RelayQueue");
 	m_SyncQueue.SetName("ApiListener, SyncQueue");
 }
 
+void ApiListener::StaticInitialize()
+{
+	int workerThreadCount = Configuration::Concurrency;
+
+	Log(LogCritical, "ApiListener")
+		<< "Concurrency for connection queues: " << workerThreadCount;
```
Comment (dnsmichi) [inline lib/remote/apilistener.cpp]: We should remove that.

**Comment by N-o-X** [comment]
Test case 1 (Master => 2500x socat):

**Setup:**
zones.conf:
```
for (i in range(2500)) {
  object Endpoint "s" + i {
     host = "192.168.10.51"
     port = "5665"
  }
  object Zone "s" + i {
    parent = ZoneName
    endpoints = [ name ]
  }
}
```

Socat:
`socat OPENSSL-LISTEN:5665,certificate=icinga-certs/sat.crt,key=icinga-certs/sat.key,cafile=icinga-certs/ca.crt,fork - > /dev/null`

Because socat doesn't send a heartbeat, its possible to just let it run into the 60 seconds timeout after connection, which allows endless testing without a restart. I've let this run for about 2 hours and didn't get one timeout. 

Connecting all 2500 endpoints took between 30 and 60 seconds on a quad core processor.

**Comment by N-o-X** [comment]
Test case 2 (Master => 2500x Firewall):

**Setup:**
zones.conf on master:
```
for (i in range(2500)) {
  object Endpoint "s" + i {
     host = "192.168.10.51"
     port = "5665"
  }
  object Zone "s" + i {
    parent = ZoneName
    endpoints = [ name ]
  }
}
```

Client:

Allow SSH
```
$ sudo ufw allow 22
```
Deny 5665
```
$ sudo ufw deny 5665
```
Enable firewall
```
$ sudo ufw enable
```

**Result:**
Icinga starts 16 threads (4 cores * 4) and tries to connect to the first 16 endpoints. All 16 endpoints get a timeout after circa 1 minute and 30 seconds. I think we should implement a quicker timeout here.

**Todo:**

- [ ] Implement timeout
- [ ] Test again

**Comment by N-o-X** [comment]
@Icebird2000 sure why not. Shouldn't be that hard to test, once my setup is in place.

**Comment by N-o-X** [comment]
Test case 2 (Master => 500x Client):

**Setup:**
In this case I used the same setup as in case 1, but replacing the socat instance with 500 Icinga instances.

**Result:**
Results didn't really change from case 1. All connections went through in 10-20 seconds.

**Comment by N-o-X** [comment]
Test case 5 (1000x REST API => Master):

**Setup:**
I used the following script to test the API:
```
for i in $(seq 1 1000); do
  curl -sSk -H 'Accept: application/json' -X PUT https://root:icinga@192.168.10.75:5665/v1/objects/hosts/$(shuf -i 1-9999999 -n 1) -d '{ "templates": [ "generic-host" ], "attrs": { "address": "192.168.1.1", "check_command": "hostalive", "vars.os" : "Linux" }, "pretty": true }' > /dev/null &
done
```

It fires 1000 parallel requests at the master, creating a new host object each time.

**Result:**
On an Icinga instance without other load, it took about 20-30 seconds to complete all requests.
I ran this test 25 times and had 7 handshake timeouts on curls side, which is pretty okay for 25000 request.

**Comment by dnsmichi** [comment]
I'm closing this in favour of #6633. The reasoning is explained in #6517.

**Pull request closed without merging**